custom_components/elasticsearch/es_doc_publisher.py

```diff
@@ -268,16 +276,33 @@ def _should_publish_entity_state(self, domain: str, entity_id: str):
     def _state_to_bulk_action(self, state: State, time: datetime):
         """Create a bulk action from the given state object."""
 
-        document = self._document_creator.state_to_document(state, time)
 
-        return {
-            "_op_type": "index",
-            "_index": self._index_alias,
-            "_source": document,
-            # If we aren't writing to an alias, that means the
-            # Index Template likely wasn't created properly, and we should bail.
-            "require_alias": True,
-        }
+        if self._destination_type == INDEX_MODE_DATASTREAM:
+            document = self._document_creator.state_to_document(state, time, version=2)
+            # <type>-<name>-<namespace>
+            # <datastream_prefix>.<domain>-<suffix>
+            # metrics-homeassistant.device_tracker-default
+            destination_data_stream = self.datastream_prefix + "." + state.domain + "-" + self.datastream_suffix
```

Comment:
I don't expect this to be a problem in practice, but since we aren't in control of `state.domain`, do you think it'd make sense to ensure that it conforms to the DS naming conventions, in terms of casing & allowed/disallowed characters?

Reply:
So it looks like the criteria for valid domain names is much more strict than the criteria for invalid datastream names. The only valid domain names are letters + numbers + underscores. I'll include a sanitize function I won't be able to wire up any real doc publisher tests for it